src/which.js

```diff
@@ -46,23 +51,32 @@ function _which(options, cmd) {
       var attempt = path.resolve(dir + '/' + cmd);
 
       if (common.platform === 'win') {
-        var baseAttempt = attempt;
-        attempt = baseAttempt + '.exe';
-        if (checkPath(attempt)) {
-          where = attempt;
-          return;
-        }
-        attempt = baseAttempt + '.bat';
-        if (checkPath(attempt)) {
-          where = attempt;
-          return;
+        // In case the PATHEXT variable is somehow not set (e.g.
+        // child_process.spawn with an empty environment), use the XP default.
+        var pathExtEnv = process.env.PATHEXT || XP_DEFAULT_PATHEXT;
+        var pathExtArray = splitPath(pathExtEnv);
+        var i;
+
+        // If the extension is already in PATHEXT, just return that.
+        for (i = 0; i < pathExtArray.length; i++) {
+          var ext = pathExtArray[i];
+          if (attempt.slice(-ext.length) === ext && checkPath(attempt)) {
```

Comment:
I prefer case insensitive comparison. There's no guarantee (as far as I can tell) that the PATHEXT values will **always** be uppercase, and it's a little excessive to call .toUpperCase() on both sides. The regex should be very simple. Whichever is easiest is fine with me though